src/promptflow/promptflow/executor/_async_nodes_scheduler.py

```diff
@@ -117,3 +126,57 @@ async def _sync_function_to_async_task(
         return await asyncio.get_running_loop().run_in_executor(
             executor, context.invoke_tool, node, f, kwargs
         )
+
+
+def signal_handler(sig, frame):
+    """
+    Start a thread to monitor coroutines after receiving signal.
+    """
+    flow_logger.info(f"Received signal {sig}({signal.Signals(sig).name}),"
+                     " start coroutint monitor thread.")
+    loop = asyncio.get_running_loop()
+    monitor = threading.Thread(target=monitor_coroutine_after_cancellation, args=(loop,))
+    monitor.start()
+    raise KeyboardInterrupt
+
+
+def monitor_coroutine_after_cancellation(loop: asyncio.AbstractEventLoop):
+    """Exit the process when all coroutines are done.
+    We add this function because if a sync tool is running in async mode,
+    the task will be cancelled after receiving SIGINT,
+    but the thread will not be terminated and blocks the program from exiting.
+    :param loop: event loop of main thread
+    :type loop: asyncio.AbstractEventLoop
+    """
+    # TODO: Use environment variable to ensure it is flow test / node test scenario
+    # to avoid unexpected exit.
+    max_wait_seconds = os.environ.get("PF_WAIT_SECONDS_AFTER_CANCELLATION", 30)
+
+    all_tasks_are_done = False
+    exceeded_wait_seconds = False
+
+    thread_start_time = time.time()
+    flow_logger.info(f"Start to monitor coroutines after cancellation, max wait seconds: {max_wait_seconds}s")
+
+    while not all_tasks_are_done and not exceeded_wait_seconds:
+        # For sync tool running in async mode, the task will be cancelled,
+        # but the thread will not be terminated, we exit the program despite of it.
+        # TODO: Detect whether there is any sync tool running in async mode,
+        # if there is none, avoid sys._exit and let the program exit gracefully.
+        all_tasks_are_done = all(task.done() for task in asyncio.all_tasks(loop))
+        if all_tasks_are_done:
+            flow_logger.info("All coroutines are done. Exiting.")
+            os._exit(0)
+
+        exceeded_wait_seconds = time.time() - thread_start_time > max_wait_seconds
+        time.sleep(1)
+
+    if exceeded_wait_seconds and not all_tasks_are_done:
+        flow_logger.info(f"Not all coroutines are done within {max_wait_seconds}s"
+                         " after cancellation. Exiting the process despite of them."
+                         " Please config the environment variable"
+                         " PF_WAIT_SECONDS_AFTER_CANCELLATION if your tool needs"
+                         " more time to clean up after cancellation.")
+        remaining_tasks = [task for task in asyncio.all_tasks(loop) if not task.done()]
+        flow_logger.info(f"Remaining tasks: {[task.get_name() for task in remaining_tasks]}")
+        os._exit(0)
```

Comment:
Is there any othey way to exit? It seems to be a non-public API.

Reply:
Good catch, changed to `sys.exit`